### ISSUE TITLE
Add method to insert an axis

### ISSUE THREAD

**Issue by jturner314** [bluss/rust-ndarray#352]
Currently, converting an `Array1` into a one-row or one-column `Array2` is quite awkward. This comes up often when doing linear algebra. The best way I've found is to use `into_shape`:

```rust
let long_array_name = array![1, 2, 3];
let col = {
    let len = long_array_name.len();
    long_array_name.into_shape((len, 1)).unwrap()
};
assert_eq!(col, array![[1], [2], [3]]);
```

I suggest adding something like `fn insert_axis(self, axis: Axis) -> ArrayBase<S, D::Larger>`, which would be equivalent to NumPy's `numpy.expand_dims(a, axis)`. This would also generalize to higher-dimensional arrays. The example would become:

```rust
let long_array_name = array![1, 2, 3];
let col = long_array_name.insert_axis(Axis(1));
assert_eq!(col, array![[1], [2], [3]]);
```

In my own code, I've implemented a custom trait on `Array1` to add `into_row` and `into_column` methods, but a more general `insert_axis` would be very useful. By the way, thanks for writing `ndarray`! I use it for a lot of my scientific code.

**Comment by bluss** [comment]
Sounds good.

1. I think into_shape is already pretty good for this but it's far from perfect.
2. This has come up recently. [into_dimensionality()](https://github.com/bluss/rust-ndarray/pull/343) is strict, but we floated the idea of a weaker version. Something that creates the extra axes so that `.into_ndim::<Ix2>()` on an 1d array makes a 2d array just like that. I guess we need two version for choosing where to insert the axes
3. We don't need a new associated type, because we already have the inverse relation:

```rust
fn insert_axis<A, D>(self_: Array<A, D::Smaller>, axis: Axis) -> Array<A, D>
```

**Comment by bluss** [comment]
That said, no reason to not add a new associated type if that's better. Then the Ix6 -> IxDyn thing is smooth, but I generally don't consider Ix6 to be an important use case.

**Comment by jturner314** [comment]
2. `.into_ndim::<Ix2>()` is an interesting idea, and it would be sufficient for the `Ix1 -> Ix2` case in combination with the `.reversed_axes()` method. (Use `arr.into_ndim::<Ix2>()` to get a row, and `arr.into_ndim::<Ix2>().reversed_axes()` to get a column.)

   When generalizing `into_ndim` to higher dimensions, I would suggest always prepending `1` to the shape (instead of appending it) for as many additional axes are necessary, e.g. `[3] -> [1, 3]`, `[3, 2] -> [1, 3, 2]`, `[5, 6] -> [1, 1, 5, 6]`. This is analogous to the behavior of broadcasting operations.

   `into_ndim` isn't quite as general as `insert_axis` (because it doesn't take an argument for where to insert the additional axis), and you would need to decide how to handle the case of the result dimension being smaller than the input. However, I think `into_ndim` would be sufficient for most cases, and it would be simpler than `insert_axis` in some cases (e.g. `arr2.into_ndim::<Ix4>()` versus `arr2.insert_axis(Axis(0)).insert_axis(Axis(0))`). Of course, there's no reason why you couldn't have both methods (and possibly some others too, such as `numpy.swapaxes`). :-)

3. Good point; I didn't consider that.

**Comment by bluss** [comment]
Fixed by #354. Will be in ndarray 0.10.8
